stactask/task.py

```diff
@@ -273,10 +279,18 @@ def cleanup_workdir(self) -> None:
             )
 
     def assign_collections(self) -> None:
-        """Assigns new collection names based on upload_options collections attribute
-        according to the first matching expression in the order they are defined."""
-        for item in self._payload["features"]:
-            if coll := utils_find_collection(self.collection_mapping, item):
+        """Assigns new collection names based on collection_matchers or the legacy
+        upload_options collections attribute according to the first matching
+        expression in the order they are defined."""
+        if self.payload.collection_matchers:
+            collection_config: Union[dict[str, str], list[dict[str, Any]]] = (
```

Comment:
With python 3.10 being the new minimum supported version (it seems like this change is a month early, but I'm not going to push back on it) `Union` and `Optional` can go the way of the dodo. Usage of `|` should be preferred; I like to enable the ruff rule to enforce this style (and a bunch of others; look at any of my recent projects to see my curated list of useful ruff rules).

Reply:
OK. I updated to use the `|` syntax in place of `Union` and `Optional`. I'll save the ruff rule updates for a later commit (lots of changes).

Reply:
ruff updates complete.